In [1]:
import pandas as pd
import numpy as np
import seaborn as sn
import matplotlib.pyplot as plt

## Load Data
### Base Path

In [2]:
base_path = "../data/Sample_Game_1/"


### Load Events

In [3]:
events = pd.read_csv(base_path + "Sample_Game_1_RawEventsData.csv")
events.head()

,Team,Type,Subtype,Period,Start Frame,Start Time [s],End Frame,End Time [s],From,To,Start X,Start Y,End X,End Y
0,Away,SET PIECE,KICK OFF,1,1,0.04,0,0.00,Player19,NaN,NaN,NaN,NaN,NaN
1,Away,PASS,NaN,1,1,0.04,3,0.12,Player19,Player21,0.45,0.39,0.55,0.43
2,Away,PASS,NaN,1,3,0.12,17,0.68,Player21,Player15,0.55,0.43,0.58,0.21
3,Away,PASS,NaN,1,45,1.80,61,2.44,Player15,Player19,0.55,0.19,0.45,0.31
4,Away,PASS,NaN,1,77,3.08,96,3.84,Player19,Player21,0.45,0.32,0.49,0.47


### Load Tracking (Home)

In [4]:
tracking_home = pd.read_csv(
    base_path + "Sample_Game_1_RawTrackingData_Home_Team.csv"
)

tracking_home.head()

C:\Users\30697\AppData\Local\Temp\ipykernel_25924\1918793147.py:1: DtypeWarning: Columns (0,1,2,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31) have mixed types. Specify dtype option on import or set low_memory=False.
  tracking_home = pd.read_csv(


,Unnamed: 0,Unnamed: 1,Unnamed: 2,Home,Unnamed: 4,Home.1,Unnamed: 6,Home.2,Unnamed: 8,Home.3,...,Home.10,Unnamed: 24,Home.11,Unnamed: 26,Home.12,Unnamed: 28,Home.13,Unnamed: 30,Unnamed: 31,Unnamed: 32
0,NaN,NaN,NaN,11,NaN,1,NaN,2,NaN,3,...,10,NaN,12,NaN,13,NaN,14,NaN,NaN,NaN
1,Period,Frame,Time [s],Player11,NaN,Player1,NaN,Player2,NaN,Player3,...,Player10,NaN,Player12,NaN,Player13,NaN,Player14,NaN,Ball,NaN
2,1,1,0.04,0.00082,0.48238,0.32648,0.65322,0.33701,0.48863,0.30927,...,0.55243,0.43269,NaN,NaN,NaN,NaN,NaN,NaN,0.45472,0.38709
3,1,2,0.08,0.00096,0.48238,0.32648,0.65322,0.33701,0.48863,0.30927,...,0.55243,0.43269,NaN,NaN,NaN,NaN,NaN,NaN,0.49645,0.40656
4,1,3,0.12,0.00114,0.48238,0.32648,0.65322,0.33701,0.48863,0.30927,...,0.55243,0.43269,NaN,NaN,NaN,NaN,NaN,NaN,0.53716,0.42556


### Load Tracking (Away)

In [5]:
tracking_away = pd.read_csv(
    base_path + "Sample_Game_1_RawTrackingData_Away_Team.csv"
)

tracking_away.head()

C:\Users\30697\AppData\Local\Temp\ipykernel_25924\3652936359.py:1: DtypeWarning: Columns (0,1,2,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31) have mixed types. Specify dtype option on import or set low_memory=False.
  tracking_away = pd.read_csv(


,Unnamed: 0,Unnamed: 1,Unnamed: 2,Away,Unnamed: 4,Away.1,Unnamed: 6,Away.2,Unnamed: 8,Away.3,...,Away.10,Unnamed: 24,Away.11,Unnamed: 26,Away.12,Unnamed: 28,Away.13,Unnamed: 30,Away.14,Unnamed: 32
0,NaN,NaN,NaN,25,NaN,15,NaN,16,NaN,17,...,24,NaN,26,NaN,27,NaN,28,NaN,NaN,NaN
1,Period,Frame,Time [s],Player25,NaN,Player15,NaN,Player16,NaN,Player17,...,Player24,NaN,Player26,NaN,Player27,NaN,Player28,NaN,Ball,NaN
2,1,1,0.04,0.90509,0.47462,0.58393,0.20794,0.67658,0.4671,0.6731,...,0.37833,0.27383,NaN,NaN,NaN,NaN,NaN,NaN,0.45472,0.38709
3,1,2,0.08,0.90494,0.47462,0.58393,0.20794,0.67658,0.4671,0.6731,...,0.37833,0.27383,NaN,NaN,NaN,NaN,NaN,NaN,0.49645,0.40656
4,1,3,0.12,0.90434,0.47463,0.58393,0.20794,0.67658,0.4671,0.6731,...,0.37833,0.27383,NaN,NaN,NaN,NaN,NaN,NaN,0.53716,0.42556


## Data Familiarization

### Shape of each dataset

In [6]:
print("Events shape:", events.shape)
print("Home tracking shape:", tracking_home.shape)
print("Away tracking shape:", tracking_away.shape)

Events shape: (1745, 14)
Home tracking shape: (145008, 33)
Away tracking shape: (145008, 33)


### Events
#### Columns & summary table

In [7]:
events.columns

Index(['Team', 'Type', 'Subtype', 'Period', 'Start Frame', 'Start Time [s]',
       'End Frame', 'End Time [s]', 'From', 'To', 'Start X', 'Start Y',
       'End X', 'End Y'],
      dtype='object')

In [8]:
events.info()
events.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1745 entries, 0 to 1744
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Team            1745 non-null   object 
 1   Type            1745 non-null   object 
 2   Subtype         792 non-null    object 
 3   Period          1745 non-null   int64  
 4   Start Frame     1745 non-null   int64  
 5   Start Time [s]  1745 non-null   float64
 6   End Frame       1745 non-null   int64  
 7   End Time [s]    1745 non-null   float64
 8   From            1745 non-null   object 
 9   To              799 non-null    object 
 10  Start X         1664 non-null   float64
 11  Start Y         1664 non-null   float64
 12  End X           1109 non-null   float64
 13  End Y           1109 non-null   float64
dtypes: float64(6), int64(3), object(5)
memory usage: 191.0+ KB


,Team,Type,Subtype,Period,Start Frame,Start Time [s],End Frame,End Time [s],From,To,Start X,Start Y,End X,End Y
count,1745,1745,792,1745.000000,1745.000000,1745.000000,1745.000000,1745.000000,1745,799,1664.000000,1664.000000,1109.000000,1109.000000
unique,2,9,51,NaN,NaN,NaN,NaN,NaN,28,28,NaN,NaN,NaN,NaN
top,Home,PASS,INTERCEPTION,NaN,NaN,NaN,NaN,NaN,Player19,Player21,NaN,NaN,NaN,NaN
freq,917,799,290,NaN,NaN,NaN,NaN,NaN,103,69,NaN,NaN,NaN,NaN
mean,NaN,NaN,NaN,1.488825,70503.382235,2820.135289,70526.925501,2821.077020,NaN,NaN,0.527139,0.502524,0.524454,0.503201
std,NaN,NaN,NaN,0.500018,41925.812208,1677.032488,41925.545286,1677.021811,NaN,NaN,0.243625,0.283635,0.242372,0.299590
min,NaN,NaN,NaN,1.000000,1.000000,0.040000,0.000000,0.000000,NaN,NaN,0.000000,-0.020000,-0.050000,-0.040000
25%,NaN,NaN,NaN,1.000000,32338.000000,1293.520000,32338.000000,1293.520000,NaN,NaN,0.360000,0.270000,0.350000,0.250000
50%,NaN,NaN,NaN,1.000000,68920.000000,2756.800000,68945.000000,2757.800000,NaN,NaN,0.530000,0.490000,0.530000,0.490000
75%,NaN,NaN,NaN,2.000000,107787.000000,4311.480000,107802.000000,4312.080000,NaN,NaN,0.702500,0.752500,0.690000,0.780000


#### Basic Info

The events dataset contains 1,745 annotated match events across two teams and two match periods.
Passes dominate the event distribution (~X%), which aligns with expected football match dynamics.
The presence of 51 unique subtypes suggests fine-grained event categorization (e.g., interceptions, clearances, etc.).
Missing values in spatial columns (Start X/Y, End X/Y) indicate that not all events are spatially defined (e.g., fouls, offsides), which will require conditional handling during spatial analysis.

#### Frames Analysis

The relationship between frame index and time usually is:

$$
\text{Time (seconds)} = \frac{\text{Frame}}{25}
$$

Equivalently:

$$
\text{Frame} = 25 \times \text{Time (seconds)}
$$

In [13]:
# Check frame theory for start and end times
events_frames=events[['Start Frame', 'End Frame','Start Time [s]','End Time [s]']].copy()
# Calculate time deriving from frame number (assuming 25 fps)
events_frames['Computed_Time_from_Frame_Start'] = events_frames['Start Frame'] / 25
events_frames['Computed_Time_from_Frame_End'] = events_frames['End Frame'] / 25

# Difference between computed time and actual start time
events_frames['Start_Time_Diff'] = (
    events_frames['Computed_Time_from_Frame_Start'] - events_frames['Start Time [s]']
)
events_frames['End_Time_Diff'] = (
    events_frames['Computed_Time_from_Frame_End'] - events_frames['End Time [s]']
)

# Max absolute difference

print("Max absolute difference:",
      events_frames['Start_Time_Diff'].abs().max()) # Should be close to zero if frame theory holds
print("Max absolute difference (End Time):",events_frames['End_Time_Diff'].abs().max())

events_frames[['Start Frame', 'Start Time [s]','End Frame', 'End Time [s]',
        'Computed_Time_from_Frame_Start', 'Computed_Time_from_Frame_End', 'Start_Time_Diff', 'End_Time_Diff']].head()

Max absolute difference: 0.0
Max absolute difference (End Time): 0.0


,Start Frame,Start Time [s],End Frame,End Time [s],Computed_Time_from_Frame_Start,Computed_Time_from_Frame_End,Start_Time_Diff,End_Time_Diff
0,1,0.04,0,0.00,0.04,0.00,0.0,0.0
1,1,0.04,3,0.12,0.04,0.12,0.0,0.0
2,3,0.12,17,0.68,0.12,0.68,0.0,0.0
3,45,1.80,61,2.44,1.80,2.44,0.0,0.0
4,77,3.08,96,3.84,3.08,3.84,0.0,0.0
